rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,636 @@
+|      integer      |      Integer      |        integer        |
+|      double       |      Number       |        double         |
+|      boolean      |      Boolean      |        boolean        |
+|       list        |       List        |   inner value type    |
+|       dict        |       Dict        |        object         |
+
+Note the date range is handled as other date ranges using the elasticsearch dumper (`EDTFListDumperExt`) to extract the gte (lower) and lte (upper) date limits.
+
+#### Data validation
+
+Custom fields data values are validated using Marshmallow. A `custom_fields` field is added to the `RecordsService`'s schema:
+
+```python
+class RecordSchema:
+     metadata = ...
+     custom_fields = fields.Nested(CustomFieldsSchema)
+```
+
+It will load from configuration all custom fields an generate a Marshmallow schema. The resulting schema will have as fields the values returned by the `<Type>CF.field` function. For example `SanitizedUnicode` in the `TextCF` presented above.
+
+```python
+class CustomFieldsSchema(Schema):
+     """Marshmallow schema for custom fields.
+
+     Loads all schemas from the configured fields.
+     Uses the singleton pattern to avoid loading multiple times.
+     """
+
+     def __init__(self, *args, **kwargs):
+         """constructor."""
+         super().__init__(*args, **kwargs)
+         config = current_app.config.get("RDM_CUSTOM_FIELDS", {})
+         self.fields = {
+             field_key: getattr(field, "field")()
+             for field_key, field in config.items()
+         }
+         self._schema = Schema.from_dict(self.fields)()
+
+     def _serialize(self, obj, **kwargs):
+         """Dumps the custom fields values."""
+         return self._schema.dump(obj)
+
+     def _deserialize(self, data, **kwargs):
+         """Loads the custom fields values."""
+         return self._schema.load(data=data)
+```
+
+### Linking custom fields to vocabularies
+
+> Note: in this sub section the term _record_ is treated as a generic term, namely a document. It will refer to bibliographic and community records as _records_.
+
+Simple custom fields types are added to the record's data model and dumped in its json blob. However, the vocabulary type works differently as it is a relation to another record, and it needs to be linked. Meaning, the selected value need to be checked for existence when a record is created and dereferenced when it is read.
+
+The programmatic API to access records' relations is kept intact. For example, a field named `myrelationfield` would be accessed as `record.relations.myrelationfield`. However, a new class `MultiRelationsField` has been implemented to accept not only implementations of `RelationBase` but also `RelationsField`.
+
+`CustomFieldsRelation` is an implementation of `RelationsField` that injects vocabulary custom fields in the record's relations attribute. In addition, to avoid name collisions these new fields are prefixed with `custom_`.
```

Comment:
This section has two issues now:

- The prefix should be `custom_fields_`
- What about namespaces?

I fear we might end up with fields like `relations.custom_fields_cern_experiment`, which is very long and not dev frienly. However, this is **only a problem in the programmatic API**, which being honest we do not use so ofter for relations... Also, I think it is **not a blocker**, and can be figured out later on. cc @zzacharo @lnielsen

Reply:
I would probably prepend the namespace, so users will know what to expect if they define long namespaces. For example, you would access it as `record.relations.cern_experiment`...